posts/frangeris/04.upgradeable-contracts.md

```diff
@@ -0,0 +1,228 @@
+
+### **How the proxy pattern works?**
+
+The secret behinds a proxy consists in a contract that provides a fallback function that delegates all calls to another contract using the EVM instruction `delegatecall`, this is a low-level function that allows calling another contract using the current contract **storage** and **context**. This means that the contract that is being called can access the storage of the contract that is calling it. This is the key to the proxy pattern.
+
+Here is an example of how `delegatecall` works:
+
+```solidity
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.17;
+
+// NOTE: Deploy this contract first
+contract B {
+    // NOTE: storage layout must be the same as contract A
+    uint public num;
+    address public sender;
+    uint public value;
+
+    function setVars(uint _num) public payable {
+        num = _num;
+        sender = msg.sender;
+        value = msg.value;
+    }
+}
+
+contract A {
+    uint public num;
+    address public sender;
+    uint public value;
+
+    function setVars(address _contract, uint _num) public payable {
+        // A's storage is update, B state is not modified.
+        (bool success, bytes memory data) = _contract.delegatecall(
+            abi.encodeWithSignature("setVars(uint256)", _num)
+        );
+    }
+}
+```
+
+In short words, the proxy contract is the one that DApps interact with, while using the logic of other contract in the proxy context. This means that the proxy contract will be the one that is deployed and the one that users will interact with, while the logic contract will be the one that is upgraded.
+
+![Meme about how proxy is behind scenes](./assets/04.image.1.png)
+
+More technical documentation can be found on the [Solidity docs](https://docs.soliditylang.org/en/v0.8.20/introduction-to-smart-contracts.html#delegatecall-and-libraries).
+
+> When working with upgradeable contracts using OpenZeppelin Upgrades, there are a few minor caveats to keep in mind when writing your Solidity code.
+>
+> It’s worth mentioning that these restrictions have their roots in how the Ethereum VM works, and apply to all projects that work with upgradeable contracts, not just OpenZeppelin Upgrades.
+
+Some of the caveats of using the proxy pattern are:
+
+1. Initializers: Contructors are not allowed in upgradeable contracts, instead we need to use the `initialize` function from the `Initializable` contract. This function is called only once when the contract is deployed for the first time. This means that we need to use the `initializer` modifier in the `initialize` function.
+
+2. Extending from upgradeable contracts libraries: Following the same logic, we need to use the `initializer` modifier when extending from upgradeable contracts libraries for libraries that we are extending from.
```

Comment:
I don't really understand the idea here. Which are the libraries "_we are extending from_"??
In any case, if you extend _from_ something, it sounds to me that the second preposition should be _to_

Reply:
A smart contract is like a `class`, you can use inheritance in the same way we do in OOP

Reply:
Alright then. So maybe this could be written this way:

```suggestion
2. Extending from upgradeable contracts libraries: Following the same logic, we need to use the `initializer` modifier when extending from upgradeable contract libraries.
```